### ISSUE TITLE
Graphql resolvers called when data already present

### ISSUE THREAD

**Issue by cjreimer** [redwoodjs/graphql#3726]
I'm working on some query optimization and noted that qraphql resolvers are all called when the data is already present.  For routinely accessed relations, we use Prisma to load the relation with the original query, as follows:

```
  const result = (await db.requirement.findMany({
    include: {
      actionBy: true,
    },
  }))
```

When putting Prisma into query logging mode, I noted that we were querying the `actionBy `relation multiple times.

Adding the following to the `actionBy `resolver confirmed this, as the console log was generated n times for each row in the original query.

```
  actionBy: (_obj, { root }) => {
    console.log('actionBy resolver')
    return db.requirement.findUnique({ where: { id: root.id } }).actionBy()
  },

```

Doing the following resolved the issue.
```
  actionBy: (_obj, { root }) => {
    if ('actionBy' in root) return root.actionBy
    return db.requirement.findUnique({ where: { id: root.id } }).actionBy()
  },

```

Shouldn't our graphql server do this check by default?  Is this a Helix issue?

@dthyresson can I tag you on this?

Thanks!  I appreciate it!

**Comment by dthyresson** [comment]
Hi @cjreimer you've stumbled upon the classic GraphQL N+1 query where (due to the nested, graph nature of a query) a query can issue many, many sql calls.

This isn't anything really Helix or Prisma can do on its own. I also find it really frustrating and do lots of logging at the Prisma and resolver level to see if there is some query "explosion" when I test locally.

This is so common that there is even a pattern to try to alleviate this behavior called Dataloader: https://www.graphql-modules.com/docs/recipes/dataloader

I haven't tried it yet but there is an envelop plugin that help: https://github.com/dotansimha/envelop/tree/main/packages/plugins/dataloader

Also, I find this useTiming envelop plugin handy: https://github.com/dotansimha/envelop/blob/main/packages/core/docs/use-timing.md

It shows me how much time each resolver is taking. So, even resolving fields ( query for name, title from post) like name and title take time to populate -- not re-query but populate it can reveal that time spent.

I assume that the shortcut of:

` if ('actionBy' in root) return root.actionBy`

only works if you use the `include` in the query -- that is, what Rails called "eager loading"?

I take it the shortcut says, if the document already has the "actionBy" field populated, don't bother fetching again, right?

**Comment by cjreimer** [comment]
Thanks, @dthyresson !  

> I take it the shortcut says, if the document already has the "actionBy" field populated, don't bother fetching again, right?

Yes, the code I added simply does not refetch the data if it already exists.

I just assumed that the graphql server could/would do this inherently behind the scenes.  However, if this is standard expected behavior, then that's fine.

I made almost an order of magnitude improvement on our query time with this optimization, so I will be looking out for this!

Feel free to close if you don't think it's worth chasing further.

**Comment by cjreimer** [comment]
@dthyresson, that sounds like a good, reasonably balanced, solution.  Adding it to the template adds very little to the codebase for the user, and if they want to get rid of the check, they easily can.   

I like your plan!

**Comment by dthyresson** [comment]
> @dthyresson, that sounds like a good, reasonably balanced, solution. Adding it to the template adds very little to the codebase for the user, and if they want to get rid of the check, they easily can.
> 
> I like your plan!

@cjreimer Are you using TypeScript by any chance?

When I try something like:

```
if ('actionBy' in root) return root.actionBy
```

In my case:

```
export const Track = {
  album: (_obj, { root }: ResolverArgs<ReturnType<typeof track>>) => {
    if ('album' in root) return root.album
    db.track.findUnique({ where: { id: root.id } }).album()
  },
```

I get a TS error:

```
Property 'album' does not exist on type 'Track'.ts(2339)
```

**Comment by cjreimer** [comment]
Yes, we are typescript.  OK, here's where I need to explain what we do a little more, as we use less of the strict Prisma types that are in the templates and either of the following:

a:  Generate a model type from the Prisma type that includes the nested objects as follows:
```
import {
  Prisma,
  ReqAction as PrismaReqAction,
  Requirement,
  Submittal,
  User,
} from '@prisma/client'

export type ReqActionModel = PrismaReqAction & {
  createdBy?: User
  updatedBy?: User
  req?: Requirement
  submittal?: Submittal
}
export const ReqAction = {
  createdBy: (_obj, { root }: ResolverArgs<ReqActionModel>) => {
    if ('createdBy' in root) return root.createdBy
    return db.reqAction.findUnique({ where: { id: root.id } }).createdBy()
  },
}

```

or... and this is what we are moving toward
b.  Import the graphQL type that is generated by redwood for us.  As this is a graphql problem, the graphQL type will more closely match the graphQL schema

```
import {
  ReqAction as ReqActionModel,
  CreateReqActionInput,
  CreateReqCommentInput,
  CreateReqSubmittalInput,
  SubmitReqSubmittalInput,
} from 'types/graphql'

export const ReqAction = {
  createdBy: (_obj, { root }: ResolverArgs<ReqActionModel>) => {
    if ('createdBy' in root) return root.createdBy
    return db.reqAction.findUnique({ where: { id: root.id } }).createdBy()
  },
}
```

Does that make sense?

Note from my import above, that we use the graphQL types as the types for the service function inputs rather than the prisma types (Another topic!).

**Comment by dac09** [comment]
It's taken me a while to fully wrap my head around this, but Curtis and DT's suggestions here make a lot of sense to me. I'm putting in the v1 priority queue for now - I think the solution seems reasonable and a good default to me.

**Comment by orta** [comment]
The described pattern above I have used also in my apps, though I was expecting to eventually move to the dataloader pattern after I found a good way to measure the number of Prisma calls a request makes. The biggest downside to the pattern is that you might make a request earlier for partial data of an object and then in another resolver make the assumption that you have a full object, but that requires folks to be really tuning their prisma calls.

**Comment by dthyresson** [comment]
@dac09 Let's pair on this one as well?

**Comment by Tobbe** [comment]
> The biggest downside to the pattern is that you might make a request earlier for partial data of an object and then in another resolver make the assumption that you have a full object

Wanted to highlight this ☝️  @cjreimer's optimization is a powerful one that can bring serious performance wins. But it can also lead to bugs that might be difficult to understand, especially if this is generated code and you're new to gql and/or prisma.

I don't know if the potential optimization is worth the risks. Might be. Maybe some clear comments in the generated code is all that's needed. Just wanted to make sure this was all clear before any decisions were made :)

**Comment by cjreimer** [comment]
Yes, I agree there are risks, but so far they have been well worth it for us!  I do agree with @Tobbe 's suggestion to add comments in the generated code.  I believe that's a reasonable balance.

There are likely some other alternatives but would be significantly more complex.  This could be a future discussion.

**Comment by orta** [comment]
FWIW: What this looks like in my codebase today:

```ts
export const Activity: ActivityQueries = {
  actor: (_obj, { root }) =>
    "actorID" !== undefined
      ? db.user.findUnique({ where: { id: root.actorID } })
      : db.activity.findUnique({ where: { id: root.id } }).actor(),
   // repeat...
```

`ActivityQueries` comes from https://github.com/redwoodjs/redwoodjs.com/issues/878#issuecomment-1037987452
